crates/chain/chain.rs

```diff
@@ -0,0 +1,185 @@
+};
+use ethereum_rust_storage::error::StoreError;
+use ethereum_rust_storage::Store;
+use thiserror::Error;
+
+#[derive(Debug, Error)]
+pub enum ChainError {
+    #[error("Invalid Block: {0}")]
+    InvalidBlock(InvalidBlockError),
+    #[error("Parent block not found (block with block_number-1)")]
+    ParentNotFound,
+    #[error("Block number is not the latest plus one")]
+    GreaterThanLatestPlusOne,
+    #[error("Previous block's hash does not match parent_hash in header")]
+    ParentHashMismatch,
+    #[error("DB error: {0}")]
+    StoreError(StoreError),
+    #[error("EVM error: {0}")]
+    EvmError(EvmError),
+}
+
+#[derive(Debug, Error)]
+pub enum InvalidBlockError {
+    #[error("World State Root does not match the one in the header after executing")]
+    StateRootMismatch,
+    #[error("Invalid Header, validation failed pre-execution")]
+    InvalidHeader,
+    #[error("Exceeded MAX_BLOB_GAS_PER_BLOCK")]
+    ExceededMaxBlobGasPerBlock,
+    #[error("Exceeded MAX_BLOB_NUMBER_PER_BLOCK")]
+    ExceededMaxBlobNumberPerBlock,
+    #[error("blob gas used doesn't match value in header")]
+    BlobGasUsedMismatch,
+}
+
+/// Adds a new block as head of the chain.
+/// Performs pre and post execution validation, and updates the database.
+pub fn add_block(block: &Block, storage: Store) -> Result<(), ChainError> {
+    // Validate if it can be the new head and find the parent
+    let parent_header = find_parent_header(block, &storage)?;
+
+    let mut state = evm_state(storage.clone());
+
+    // Validate the block pre-execution
+    validate_block(block, &parent_header, &state)?;
+
+    execute_block(block, &mut state).map_err(|e| ChainError::EvmError(e))?;
+
+    apply_state_transitions(&mut state).map_err(|e| ChainError::StoreError(e))?;
+
+    // Check state root matches the one in block header after execution
+    validate_state(&block.header, storage.clone())?;
+
+    store_block(storage.clone(), block.clone())?;
+
+    Ok(())
+}
+
+/// Stores block and header in the database
+pub fn store_block(storage: Store, block: Block) -> Result<(), ChainError> {
+    storage
+        .add_block(block.clone())
+        .map_err(|e| ChainError::StoreError(e))?;
+    storage
+        .update_latest_block_number(block.header.number)
+        .map_err(|e| ChainError::StoreError(e))?;
+    Ok(())
+}
+
+/// Performs post-execution checks
+pub fn validate_state(block_header: &BlockHeader, storage: Store) -> Result<(), ChainError> {
+    // Compare state root
+    if storage.world_state_root() == block_header.state_root {
+        Ok(())
+    } else {
+        return Err(ChainError::InvalidBlock(
+            InvalidBlockError::StateRootMismatch,
+        ));
+    }
+}
+
+/// Validates if the provided block could be the new head of the chain, and returns the
+/// parent_header in that case.
+fn find_parent_header(block: &Block, storage: &Store) -> Result<BlockHeader, ChainError> {
```

Comment:
This function should look for the parent block based on the hash value of the parent, not on the height of the block.

Reply:
That makes sense, if we can find the parent by hash then the block can be valid even if the block number is not the latest plus one, it could be a block we already have on the chain or we could be performing a reorg. 

But I think fetching and storing blocks by block_hash is not supported by our store yet, only by block number, any suggestions on how to proceed?

I could iterate through the blocks by block number and check if the hash matches until we have that feature for the store or I could try to implement it myself.

Reply:
https://github.com/lambdaclass/ethereum_rust/issues/280
I opened an issue for fetching blocks by hash.